### PR TITLE
Command history

### PR DESCRIPTION
**Pull request by mjgpy3**
History + the feature we were seeing yesterday (navigate to an empty command). This probably isn't ready to be merged yet as I realized there's nothing handling adding "unknown commands" to CommandHistory.

### CODE REVIEW

File: src/Void.Core/CommandMode.fs [hunk 1]
```diff
@@ -8,13 +8,14 @@ module CommandMode =
...
+        | CommandCompleted of string
         interface EventMessage
 
     let private handleEnter interpret buffer =
         match interpret { Language = "VoidScript"; Fragment = buffer} with
         | InterpretScriptFragmentResponse.Completed ->
-            ("", Event.CommandCompleted :> Message)
+            ("", Event.CommandCompleted buffer :> Message)
         | InterpretScriptFragmentResponse.ParseFailed error ->
             ("", CoreEvent.ErrorOccurred error :> Message)
```
Comment (mjgpy3) [inline src/Void.Core/CommandMode.fs:20]: I've gotta figure out how to get these to CommandHistory and strip out the attempted command (at least for unknown command errors)

Comment (Kazark) [inline src/Void.Core/CommandMode.fs:20]: @mjgpy3 For the error, seems like you could add the command that failed as information to the error type that says that command failed, and then in your command history add a core event handler that pattern matches for `ErrorOccurred` only for `ParseFailed` errors. Or something to that effect... the fact that the error is already pre-built, not build in the command mode service, might make things a little messier. Might be in for a bit of a refactor on that.

**Comment by mjgpy3** [comment]
@Kazark, what about something like this?

```
`ModeService` -UnknownCommandEntered text---> `ModeService`
                                          +-> `CommandHistory`

`ModeService` would respond by changing to Normal Mode
`CommandHistory` would save off the text
```

**Comment by Kazark** [comment]
@mjgpy3 Well, this is not just an unknown command; it includes all parse errors. Have you looked at the `Error` type and `Errors` module in `UserNotificationTypes.fs`? The current pattern is, if there is an error, publish one of those. That might be a pattern that needs to change, but if so we need to understand that we are changing it, and why. Or is it better to follow the pattern? For example, what if the parser returned not only a string description of what failed, but also the string it was parsing when it failed?

**Comment by Kazark** [comment]
Also, the stuff with switching to Normal mode is already implemented in `ModeService`. If you think that's the wrong place, we can discuss moving it to `CommandService`. Honestly, though, it doesn't feel like that should be impacted by command history stuff.

**Comment by mjgpy3** [comment]
I see your point. I'll dwell on it some more tonight :)

**Comment by Kazark** [comment]
Yeah, by the way, how much exposure have you had to event-driven architectures? I might be jumping into the middle of things here assuming we both have the same understanding of command and event messages. We should probably step back and make sure we are philosophically aligned before discussing implementation details. My bad.

**Comment by Kazark** [comment]
Overall this PR looks great. So glad to have you on the project!!

File: src/Void.Core/CoreMessages.fs [hunk 2]
```diff
@@ -43,6 +43,20 @@ type CoreCommand =
...
+type CommandHistoryCommand =
+    | MoveToPreviousCommand
+    | MoveToNextCommand
+    interface CommandMessage
+
+
+[<RequireQualifiedAccess>]
+type CommandHistoryEvent =
+    | MovedToCommand of string
+    | MovedToEmptyCommand
```
Comment (Kazark) [inline src/Void.Core/CoreMessages.fs:56]: I like that you didn't do `MovedToCommand ""`. That would seem like a big temptation but the semantics of events are very important, and it is a different occurrence when you move out of the history list then when you are moving around in it.

Comment (mjgpy3) [inline src/Void.Core/CoreMessages.fs:56]: Cool. It seemed more "honest" to publish the separate event.

**Comment by mjgpy3** [comment]
I had the wrong type in a test. Fixed it, however, I can't seem to run tests on my machine. Are there magic steps to get those running?

**Comment by mjgpy3** [comment]
> Ah, so that's what broke the build. "I don't always compile my code, but when I do, I compile it on the build server." :stuck_out_tongue_winking_eye:

Haha, exactly!

**Comment by Kazark** [comment]
You will need to install some sort of NUnit runner. I think there is a plugin for Visual Studio. There are also standalone GUI and console runners for it.

**Comment by Kazark** [comment]
So is this ready to merge, then?

**Comment by mjgpy3** [comment]
As far as I'm concerned, yes. There will need to be later issues for things like "persistent history"

**Comment by Kazark** [comment]
I'm happy to merge this now. My only question is&mdash;and this is a question, not a requirement&mdash;do you care to add any tests before I merge merge merge?

**Comment by mjgpy3** [comment]
I think I would like to add some

**Comment by Kazark** [comment]
Okay. Let me know if you need help with that.

**Comment by mjgpy3** [comment]
Alright, will do!

**Comment by Kazark** [comment]
I think that is good. I have been relying a lot on manually regressioning stuff before merging, but that is not scalable.

**Comment by mjgpy3** [comment]
I added some tests, nothing exhaustive. These are very very open to suggestions. They feel kind of bulky for testing so little.

**Comment by mjgpy3** [comment]
I like it! I was trying to avoid modifying the internal history state, however, I guess knowledge of it is inherently external (a la `CommandHistory.Service`). I'll give the "modify history" approach a shot!

I am using NCrunch. I like it so far. I'm having some pain seeing why things are failing but I think that's an NUnit thing.

**Comment by Kazark** [comment]
Haha, that's the way! Think like Git: just rebase everything! ;)

Yeah, actually one of the things that drove me to design the services the way I did was for testability. One of the things Joel taught me was that the only way to make a state machine testable is to inject the state.

I do want other services to treat the state as private, but I don't think it's too much of a liability to make the unit tests depend on the details a little bit. At least at that point you have at max two places to update something, rather than having it scattered throughout the code.

**Comment by Kazark** [comment]
Looks good. Thanks so much, Mike! This resolves #35

**Comment by mjgpy3** [comment]
Thank you! This was fun. I'll have to hop on another one soon :+1:

**Pull request merged**